### DISCUSSION TITLE
Certificate sign fails due to OSSL_SIGNATURE_PARAM_ALGORITHM_ID missing?

### DISCUSSION THREAD

**Discussion by commel** [openssl/openssl#24606] [Q&A] [answered]
My custom provider uses an ED25519 key with a custom OID. Loading the keys work fine, but when I try to create a CSR the sign fails at Line 183:

https://github.com/openssl/openssl/blob/1977c00f00ad0546421a5ec0b40c1326aee4cddb/crypto/asn1/a_sign.c#L166-L190

The key loader looks like this

```c
static void *keymgmt_load(const void *reference, size_t reference_sz) 
{
  EVP_PKEY* pkey = EVP_PKEY_new_raw_private_key(EVP_PKEY_ED25519, NULL, (const unsigned char*) reference, reference_sz);
  return pkey;
}
```

I wonder why my Key does not contain a OSSL_SIGNATURE_PARAM_ALGORITHM_ID param?

**Comment by t8m** [comment, marked as answer]
Your provider must provide such parameter on the get_ctx_params() call on the signature context.

**Reply by t8m** [reply]
See https://github.com/openssl/openssl/blob/master/providers/implementations/signature/eddsa_sig.c#L493-L496

**Reply by commel** [reply]
Thank you! This helped! Just had to wrap my head around not to pass this information but to present it in the get_ctx_params-function to be requested.

Also this commented helped a lot, as I had to encode my custom OID into the X509_ALGOR-struct: https://github.com/openssl/openssl/issues/22932#issuecomment-1846707212
